ktor-server/ktor-server-cio/jvmAndNix/src/io/ktor/server/cio/CIOApplicationEngine.kt

```diff
@@ -113,6 +117,51 @@ public class CIOApplicationEngine(
         }
     }
 
+    private suspend fun addHandlerForExpectedHeader(output: ByteWriteChannel, call: CIOApplicationCall) {
+        val continueResponse = "HTTP/1.1 100 Continue\r\n"
+        val expectHeaderValue = "100-continue"
+
+        val expectHeader = call.request.headers[HttpHeaders.Expect]?.lowercase()
+        var closeConnection = true
+
+        val expectedHeaderPhase = PipelinePhase("ExpectedHeaderPhase")
+        call.request.pipeline.insertPhaseBefore(ApplicationReceivePipeline.Before, expectedHeaderPhase)
+        call.request.pipeline.intercept(expectedHeaderPhase) {
+            val request = call.request
+            val version = HttpProtocolVersion.parse(request.httpVersion)
+            val hasBody = hasBody(request)
+
+            if (expectHeader == null || version == HttpProtocolVersion.HTTP_1_0 || !hasBody) {
+                return@intercept
+            }
+
+            if (expectHeader != expectHeaderValue) {
+                call.respond(HttpStatusCode.ExpectationFailed)
+            } else {
+                output.apply {
+                    output.writeStringUtf8(continueResponse)
+                    output.flush()
+                }
+                closeConnection = false
+            }
+        }
+
+        call.response.pipeline.intercept(ApplicationSendPipeline.Transform) {
+            if (expectHeader == null) {
+                return@intercept
+            }
+
+            val connection = if (closeConnection) "close" else "keep-alive"
+            call.response.headers.append(HttpHeaders.Connection, connection)
```

Comment:
why do you need to modify the connection header here?

Reply:
I tried to implement the next case:

> A server that responds with a final status code before reading the
entire message body SHOULD indicate in that response whether it
intends to close the connection or continue reading and discarding
the request message (see [Section 6.6 of [RFC7230]](https://www.rfc-editor.org/rfc/rfc7230#section-6.6)).

Reply:
I think it's redundant here, because handling of connection will happen anyway in the call handling.